devices/ePaper/Drivers/Gdew0154x/Gdew0154x.cs

```diff
@@ -0,0 +1,636 @@
+
+        /// <summary>
+        /// Gets the index of the first frame page.
+        /// </summary>
+        protected int FirstPageIndex { get; } = 0;
+
+        /// <inheritdoc/>
+        public virtual void BeginFrameDraw()
+        {
+            // make sure we start from the first page with clear buffers
+            SetFrameBufferPage(FirstPageIndex);
+        }
+
+        /// <summary>
+        /// Calculates the upper and lower bounds of the current frame buffer page.
+        /// </summary>
+        protected virtual void CalculateFrameBufferPageBounds()
+        {
+            CurrentFrameBufferPageLowerBound = CurrentFrameBufferPage * FrameBuffer.BufferByteCount / FrameBuffer.BitDepth;
+            CurrentFrameBufferPageUpperBound = (CurrentFrameBufferPage + 1) * FrameBuffer.BufferByteCount / FrameBuffer.BitDepth;
+
+            CurrentFrameBufferStartXPosition = GetXPositionFromFrameBufferIndex(CurrentFrameBufferPageLowerBound);
+            CurrentFrameBufferStartYPosition = GetYPositionFromFrameBufferIndex(CurrentFrameBufferPageLowerBound);
+
+            FrameBuffer.StartPoint = new Point(CurrentFrameBufferStartXPosition, CurrentFrameBufferStartYPosition);
+            FrameBuffer.CurrentFramePage = CurrentFrameBufferPage;
+        }
+
+        /// <inheritdoc/>
+        public virtual void Clear(bool triggerPageRefresh = false)
+        {
+            SetFrameBufferPage(FirstPageIndex);
+
+            // paging is enabled, flush as per number of pages to ensure all display RAM is cleared
+            if (PagedFrameDrawEnabled)
+            {
+                do
+                {
+                    Flush();
+
+                    // sleep for 20ms to allow other threads to have their chance to execute
+                    Thread.Sleep(20);
+
+                    CurrentFrameBufferPage++;
+                    CalculateFrameBufferPageBounds();
+                }
+                while (CurrentFrameBufferPage < PagesPerFrame);
+
+                CurrentFrameBufferPage = FirstPageIndex;
+                CalculateFrameBufferPageBounds();
+            }
+            else
+            {
+                // paging is disabled, so the internal frame covers the entire display frame. we only need to flush once.
+                Flush();
+            }
+
+            if (triggerPageRefresh)
+            {
+                PerformFullRefresh();
+            }
+        }
+
+        /// <summary>
+        /// Draws the specified buffer directly to the Black/White RAM on the display.
+        /// Call <see cref="PerformFullRefresh"/> after to update the display.
+        /// </summary>
+        /// <param name="buffer">The buffer array to draw.</param>
+        public void DirectDrawBuffer(params byte[] buffer)
+        {
+            SendCommand((byte)Command.DataStartTransmission2);
+            SendData(buffer);
+            WaitMs(5);
+
+            SendCommand((byte)Command.DisplayRefresh);
+            WaitMs(500);
```

Comment:
can you add a little comment on why to wait 5 ms and then 500 ms? Just a comment like "// 500 ms needed as per documentation page 5" will do it. So it makes it easier when maintaining and debugging all this

Reply:
I did, but I think there are some non-sense between samples I found on M5 repo and the one provided by the screen manufacturer...